File: 알고리즘/과제/11/sort_02_2019180009.py
class City:
  def __init__(self, name, x, y):
    self.name = name
    self.x, self.y = x, y
  def __repr__(self):
    return '%s(%3d,%3d)' % (self.name, self.x, self.y)
  def __lt__(self,city):
      return self.name < city.name

cities = [
  City("Links", 1336, 536),  City("Prosy", 977, 860),
  City("Rabbi", 6, 758),     City("Addle", 222, 261),
  City("Smell", 1494, 836),  City("Quite", 905, 345),
  City("Lives", 72, 714),    City("Cross", 23, 680),
  City("Synth", 1529, 785),  City("Tweak", 1046, 426),
  City("Medic", 1485, 514),  City("Glade", 660, 476),
  City("Breve", 1586, 448),  City("Hotel", 1269, 576),
  City("Toing", 398, 561),   City("Scorn", 617, 373),
  City("Imply", 1253, 403),  City("Zilch", 1289, 29),
  City("React", 296, 659),   City("Fiche", 787, 278),
]

# DSU Decorate-Sort-Undercorate
# pairs = sorted([(c.name,c) for c in cities])
# sorted_cities = [pair[1] for pair in sorted([(c.name,c) for c in cities])]

# def fff(city):
#     return city.name
# cities.sort(key = fff)

cities.sort(key = lambda c : c.name)
print(cities) # name 날리고 city만 취함

# from operator import attrgetter
# cities.sort(key = attrgetter('name'))

cities.sort(key = lambda c : c.x)
print(cities) # name 날리고 city만 취함

File: 알고리즘/과제/11/sort_02_2019180009.c
//#include <stdio.h>
//#include <stdlib.h>
//#include <string.h>
//
//typedef struct {
//    const char* name;
//    int x, y;
//} City;
//
//
//
//City cities[] = {
//    {"Links", 1336, 536}, {"Prosy", 977, 860},  {"Rabbi", 6, 758},    {"Addle", 222, 261},
//    {"Smell", 1494, 836}, {"Quite", 905, 345},  {"Lives", 72, 714},   {"Cross", 23, 680},
//    {"Synth", 1529, 785}, {"Tweak", 1046, 426}, {"Medic", 1485, 514}, {"Glade", 660, 476},
//    {"Breve", 1586, 448}, {"Hotel", 1269, 576}, {"Toing", 398, 561},  {"Scorn", 617, 373},
//    {"Imply", 1253, 403}, {"Zilch", 1289, 29},  {"React", 296, 659},  {"Fiche", 787, 278},
//};
//
//void City_print(City* c)
//{
//    printf("%s(%d,%d)", c->name, c->x, c->y);
//}
//
//
//void City_printAll(City* p, int count)
//{
//    for (int i = 0; i < count; i++, p++) {
//        City_print(p);
//        putchar(' ');
//    }
//    putchar('\n');
//}
//
//int city_compare_name(const void* a, const void* b)
//{
//    City* c1 = (City*)a;
//    City* c2 = (City*)b;
//
//    return strcmp(c1->name, c2->name);
//}
//
//int city_compare_y(const void* a, const void* b)
//{
//    City* c1 = (City *)a;
//    City* c2 = (City *)b;
//
//    return c1->y - c2->y;
//}
//
//int main(void)
//{
//    City_printAll(cities, sizeof(cities) / sizeof(cities[0]));
//    // sort here by name
//    City_printAll(cities, sizeof(cities) / sizeof(cities[0]));
//    qsort(cities, sizeof(cities) / sizeof(cities[0]), sizeof(cities[0]), city_compare_y);
//    City_printAll(cities, sizeof(cities) / sizeof(cities[0]));
//    qsort(cities, sizeof(cities) / sizeof(cities[0]), sizeof(cities[0]), city_compare_name);
//    City_printAll(cities, sizeof(cities) / sizeof(cities[0]));
//
//    return 0;
//}